File: Serenity.Script.UI/Editor/EmailEditor.cs
using jQueryApi;
using System;
using System.ComponentModel;
using System.Html;
using System.Runtime.CompilerServices;
using System.Text.RegularExpressions;

namespace Serenity
{
    using jQueryApi.UI;

    [Editor, DisplayName("E-posta")]
    [Element("<input type=\"text\"/>")]
    public class EmailEditor : Widget<EmailEditorOptions>, IStringValue, IReadOnly
    {
        public EmailEditor(jQueryObject input, EmailEditorOptions opt)
            : base(input, opt)
        {
            RegisterValidationMethods();

            input.AddClass("emailuser").RemoveClass("flexify");

            var spanAt = J("<span/>")
                .Text("@")
                .AddClass("emailat")
                .InsertAfter(input);

            var domain = J("<input type=\"text\"/>")
                .AddClass("emaildomain")
                .AddClass("flexify")
                .InsertAfter(spanAt);

            domain.Bind("blur." + this.uniqueName, delegate
            {
                var validator = domain.Closest("form").GetDataValue("validator").As<jQueryValidator>();
                if (validator != null)
                    validator.ValidateElement(input[0]);
            });

            if (!options.Domain.IsEmptyOrNull())
                domain.Value(options.Domain);

            if (options.ReadOnlyDomain)
                domain.Attribute("readonly", "readonly").AddClass("disabled").Attribute("tabindex", "-1");

            input.Bind("keypress." + this.uniqueName, delegate(jQueryEvent e)
            {
                if (e.Which == (int)'@')
                {
                    e.PreventDefault();

                    if (!options.ReadOnlyDomain)
                    {
                        domain.Focus();
                        domain.Select();
                    }
                }
            });

            domain.Bind("keypress." + this.uniqueName, delegate(jQueryEvent e)
            {
                if (e.Which == (int)'@')
                {
                    e.PreventDefault();
                }
            });

            if (!options.ReadOnlyDomain)
            {
                input.Change(e =>
                {
                    this.Value = input.GetValue();
                });
            }
        }

        [InlineCode("this.optional({element})")]
        private static bool ValidatorIsOptional(Element element)
        {
            return false;
        }

        public static void RegisterValidationMethods()
        {
            if (jQueryValidator.Methods["emailuser"] == null)
                jQueryValidator.AddMethod("emailuser", (value, element) =>
                {
                    var domain = J(element).NextAll(".emaildomain");
                    if (domain.Length > 0 && domain.GetAttribute("readonly") == null)
                    {
                        if (ValidatorIsOptional(element) &&
                            ValidatorIsOptional(domain[0]))
                            return true;

                        value = value + "@" + domain.GetValue();

                        if (Q.Config.EmailAllowOnlyAscii)
                            return new Regex(
                                @"^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$")
                                .Test(value);

                        return new Regex(@"^((([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+(\.([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+)*)|((\x22)((((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(([\x01-\x08\x0b\x0c\x0e-\x1f\x7f]|\x21|[\x23-\x5b]|[\x5d-\x7e]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(\\([\x01-\x09\x0b\x0c\x0d-\x7f]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]))))*(((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(\x22)))@((([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.)+(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))$", "i")
                            .Test(value);
                    }
                    else
                    {
                        if (Q.Config.EmailAllowOnlyAscii)
                            return ValidatorIsOptional(element) ||
                                   new Regex(@"^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+$").Test(value);

                        return ValidatorIsOptional(element) ||
                            new Regex(@"^((([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+(\.([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+)*)|((\x22)((((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(([\x01-\x08\x0b\x0c\x0e-\x1f\x7f]|\x21|[\x23-\x5b]|[\x5d-\x7e]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(\\([\x01-\x09\x0b\x0c\x0d-\x7f]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]))))*(((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(\x22)))$", "i").Test(value);
                    }
                }, "Lütfen geçerli bir e-posta adresi giriniz!");
        }

        public string Value
        {
            get 
            {
                var domain = this.element.NextAll(".emaildomain");
                var value = this.element.GetValue();
                var domainValue = domain.GetValue();
                if (value.IsEmptyOrNull())
                {
                    if (options.ReadOnlyDomain || domainValue.IsEmptyOrNull())
                        return "";

                    return "@" + domainValue;
                }

                return value + "@" + domainValue;
            }
            set 
            {
                var domain = this.element.NextAll(".emaildomain");
                value = value.TrimToNull();
                if (value == null)
                {
                    if (!options.ReadOnlyDomain)
                        domain.Value("");

                    element.Value("");
                }
                else
                {
                    var parts = value.Split("@");
                    if (parts.Length > 1)
                    {
                        if (!options.ReadOnlyDomain)
                        {
                            domain.Value(parts[1]);
                            element.Value(parts[0]);
                        }
                        else if (!options.Domain.IsEmptyOrNull())
                        {
                            if (parts[1] != options.Domain)
                                element.Value(value); // hatayı görsünler!
                            else
                                element.Value(parts[0]);
                        }
                        else
                            element.Value(parts[0]);
                    }
                    else
                        element.Value(parts[0]);
                }
            }
        }

        public bool ReadOnly
        {
            get
            {
                var domain = element.NextAll(".emaildomain");
                return !(element.GetAttribute("readonly") == null && (!options.ReadOnlyDomain || domain.GetAttribute("readonly") == null));
            }
            set
            {
                var domain = element.NextAll(".emaildomain");
                if (value)
                {
                    element.Attribute("readonly", "readonly").AddClass("readonly");
                    if (!options.ReadOnlyDomain)
                        domain.Attribute("readonly", "readonly").AddClass("readonly");
                }
                else
                {
                    element.RemoveAttr("readonly").RemoveClass("readonly");
                    if (!options.ReadOnlyDomain)
                    {
                        domain.RemoveAttr("readonly").RemoveClass("readonly");
                    }
                }
            }
        }
    }

    [Serializable, Reflectable]
    public class EmailEditorOptions
    {
        [DisplayName("Etki Alanı")]
        public string Domain { get; set; }
        [DisplayName("Etki Alanı Salt Okunur")]
        public bool ReadOnlyDomain { get; set; }
    }
}

File: packages/corelib/src/ui/editors/emaileditor.ts
import { Fluent, ValidatableElement, Validator, tryGetText } from "../../base";
import { IReadOnly, IStringValue } from "../../interfaces";
import { ValidationHelper } from "../../q";
import { Decorators } from "../../types/decorators";
import { EditorProps, EditorWidget } from "../widgets/widget";

export interface EmailEditorOptions {
    domain?: string;
    readOnlyDomain?: boolean;
}

@Decorators.registerEditor('Serenity.EmailEditor', [IStringValue, IReadOnly])
export class EmailEditor<P extends EmailEditorOptions = EmailEditorOptions> extends EditorWidget<P> {

    static override createDefaultElement() { return Fluent("input").attr("type", "text").getNode(); }   
    declare readonly domNode: HTMLInputElement;
    private readonly domain: HTMLInputElement;

    constructor(props: EditorProps<P>) {
        super(props);
        EmailEditor.registerValidationMethods();

        this.domNode.classList.add('emailuser');

        var spanAt = Fluent("span").text('@').class('emailat').insertAfter(this.domNode);

        var domain = Fluent("input").attr("type", "text").class('emaildomain').insertAfter(spanAt);
        domain.on('blur.' + this.uniqueName, function () {
            ValidationHelper.validateElement(this.domNode);
        });
        this.domain = domain.getNode();

        if (this.options.domain) {
            domain.val(this.options.domain);
        }

        if (this.options.readOnlyDomain) {
            domain.attr('readonly', 'readonly').addClass('disabled').attr('tabindex', '-1');
        }

        this.element.on('keypress.' + this.uniqueName, (e: KeyboardEvent) => {
            if (e.key === "@") {
                e.preventDefault();
                if (!this.options.readOnlyDomain) {
                    domain.getNode().focus();
                    domain.getNode().select();
                }
            }
        });

        domain.on('keypress.' + this.uniqueName, function (e: KeyboardEvent) {
            if (e.key === "@")
                e.preventDefault();
        });

        if (!this.options.readOnlyDomain) {
            Fluent.on(this.domNode, "change", e => this.set_value(this.domNode.value));
        }
    }

    static registerValidationMethods(): void {

        Validator.addMethod('emailuser', function (value, element) {

            var domain = Fluent(element).nextSibling(".emaildomain").getNode();
            if (domain && domain.getAttribute('readonly') == null) {

                if (Validator.optional(element) && Validator.optional(domain as ValidatableElement)) {
                    return true;
                }

                return Validator.methods.email(value + '@' + (domain as any).value, element);
            }
            else {
                return Validator.methods.email(value + '@dummy.com', element);
            }
        }, tryGetText("Validation.Email"));
    }

    get_value(): string {
        var value = this.domNode.value;
        var domainValue = this.domain.value;
        if (!value) {
            if (this.options.readOnlyDomain || !domainValue) {
                return '';
            }
            return '@' + domainValue;
        }
        return value + '@' + domainValue;
    }

    get value(): string {
        return this.get_value();
    }

    set_value(value: string): void {
        value = value?.trim();
        if (!value) {
            if (!this.options.readOnlyDomain)
                this.domain.value = '';
            this.domNode.value = "";
        }
        else {
            var parts = value.split('@');
            if (parts.length > 1) {
                if (!this.options.readOnlyDomain) {
                    this.domain.value = parts[1];
                    this.domNode.value = parts[0];
                }
                else if (this.options.domain) {
                    if (parts[1] !== this.options.domain)
                        this.domNode.value = value;
                    else
                        this.domNode.value = parts[0];
                }
                else
                    this.domNode.value = parts[0];
            }
            else
                this.domNode.value = parts[0];
        }
    }

    set value(v: string) {
        this.set_value(v);
    }

    get_readOnly(): boolean {
        return !(this.domNode.getAttribute("readonly") == null &&
            (!this.options.readOnlyDomain || this.domain.getAttribute('readonly') == null));
    }

    set_readOnly(value: boolean): void {
        if (value) {
            this.domNode.setAttribute('readonly', 'readonly');
            this.domNode.classList.add('readonly');
            if (!this.options.readOnlyDomain) {
                this.domain.setAttribute('readonly', 'readonly');
                this.domain.classList.add('readonly');
            }
        }
        else {
            this.domNode.removeAttribute('readonly');
            this.domNode.classList.remove('readonly');
            if (!this.options.readOnlyDomain) {
                this.domain.removeAttribute('readonly');
                this.domain.classList.remove('readonly');
            }
        }
    }
}